t/import_csv_file.t

```diff
@@ -0,0 +1,213 @@
+        # note that the case where $item2 is a bag/set is not (yet?) supported
+        $result = [];
+        my $iterator = each_array(@$item1, @$item2);
+        while ( my ($value1, $value2) = $iterator->() ) {
+            push @$result, deep_extract_keys_from($value1, $value2);
+        }
+    }
+    # Note: if $item1 and $item2 does not have same type, 
+    # leave item1 as is, deep comparison will fail
+    return $result
+}
+
+
+# Testing import of a csv file
+{
+    my $import_module = new Test::MockModule('ProductOpener::Import');
+    # mock download image to fetch image in inputs_dir
+    $import_module->mock('download_image', \&fake_download_image);
+    # inputs
+    my $my_excel = $inputs_dir . "test.xlsx";
+    my $columns_fields_json = $inputs_dir . "test.columns_fields.json";
+    my $expected_products_path = $expected_dir . "test_products.perl";
+    # clean data
+    remove_all_products();
+    # step1: parse xls
+    my ($out, $err, $csv_result) = capture (sub {
+        return scalar load_csv_or_excel_file($my_excel);
+    });
+    ok( !$csv_result->{error} );
+    # step2: get columns match
+    my $default_values_ref = {lc => "en", countries => "en"};
+    # this is the file we need
+    my $columns_fields_file = $outputs_dir . "test.columns_fields.sto";
+    create_sto_from_json($columns_fields_json, $columns_fields_file);
+    # step3 convert file
+    my $converted_file = $outputs_dir . "test.converted.csv";
+    my $conv_result;
+    ($out, $err, $conv_result) = capture (sub {
+        return scalar convert_file(
+            $default_values_ref, $my_excel, $columns_fields_file, $converted_file
+        );
+    });
+    ok( !$conv_result->{error} );
+    # step4 import file
+    my $datestring = localtime();
+    my $args = {
+        "user_id" => "test-user",
+        "org_id" => "test-org",
+        "owner_id" => "org-test-org",
+        "csv_file" => $converted_file,
+        "exported_t" => $datestring,
+    };
+    # run
+    ($out, $err) = capture (sub {
+        ProductOpener::Import::import_csv_file($args);
+    });
+    # get all products in db, sorted by code for perdictability
+    my $cursor = execute_query(sub {
+		return get_products_collection()->query({})->sort({code =>1});
+	});
+    my @products = ();
+    while ( my $doc = $cursor->next ) {
+        push(@products, $doc);
+    }
+    # expected values
+    ## no critic (ProhibitStringyEval)
+    my $expected_products = eval(path($expected_products_path)->slurp_raw());
+    ## use critic
+    # same number of products
+    is(scalar @products, scalar @$expected_products, , "As much expected produts as in db");
+    my $items_iter = each_array(@products, @$expected_products);
+    while ( my ($product, $expected_product) = $items_iter->() )
+    {
+        # load from sto
+        my $product_obj = retrieve_product($product->{code});
+        # values are those expected
+        my $cmp_product = deep_extract_keys_from($product, $expected_product);
```

Comment:
Why do we need to filter the keys? Can't we just compare the product to the expected product directly? Is it because there are a few fields that change over time? (e.g. some timestamps?)

Reply:
This is because
* there are some fields like timestamps and others that may change
* I don't want the test to break too often because of some new computation, or change in computation, etc. So I just verify the output of the code I test (not of other parts). (Of course you may also argue that integration test should look at unexpected changes, but I think maintainability counts also).

Reply:
ok, fine with me.